File: backends/quarkus-java/src/main/java/dev/stackverse/backend/BookmarkService.java
package dev.stackverse.backend;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.quarkus.security.identity.SecurityIdentity;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.ws.rs.core.Response;
import java.net.URI;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import javax.sql.DataSource;
import org.eclipse.microprofile.jwt.JsonWebToken;

@ApplicationScoped
public class BookmarkService extends ServiceSupport {
    @Inject
    public BookmarkService(
            DataSource dataSource,
            JsonWebToken jwt,
            SecurityIdentity securityIdentity,
            ObjectMapper mapper,
            Localizer localizer) {
        super(dataSource, jwt, securityIdentity, mapper, localizer);
    }

    public Response listBookmarksV1(RequestContext request) {
        int page = pagingPage(request);
        int size = pageSize(request);
        SqlWhere query = parseBookmarkListQuery(request);
        Page<Bookmark> result =
                withConnection(
                        connection -> {
                            long total =
                                    scalarLong(
                                            connection,
                                            "select count(*) from bookmarks " + query.sql(),
                                            query.params());
                            List<Object> params = new ArrayList<>(query.params());
                            params.add(size);
                            params.add(offset(page, size));
                            List<Bookmark> items =
                                    query(
                                            connection,
                                            "select * from bookmarks "
                                                    + query.sql()
                                                    + " order by created_at desc, id desc limit ? offset ?",
                                            params,
                                            ServiceSupport::bookmark);
                            return new Page<>(items, page, size, total);
                        });
        List<BookmarkResponse> items =
                result.items().stream().map(ServiceSupport::bookmarkResponse).toList();
        return v1BookmarksDeprecationHeaders(
                Response.ok(pageResponse(items, result.page(), result.size(), result.totalItems()))
                        .build());
    }

    public Response listBookmarksV2(RequestContext request) {
        int size = pageSize(request);
        SqlWhere query = parseBookmarkListQuery(request);
        Cursor cursor =
                Optional.ofNullable(singleParam(request, "cursor"))
                        .filter(value -> !value.isBlank())
                        .map(Cursor::decode)
                        .orElse(null);
        List<Bookmark> fetched =
                withConnection(
                        connection -> {
                            String where = query.sql();
                            List<Object> params = new ArrayList<>(query.params());
                            if (cursor != null) {
                                where += " and (created_at, id) < (?, ?)";
                                params.add(cursor.createdAt());
                                params.add(cursor.id());
                            }
                            params.add(size + 1);
                            return query(
                                    connection,
                                    "select * from bookmarks "
                                            + where
                                            + " order by created_at desc, id desc limit ?",
                                    params,
                                    ServiceSupport::bookmark);
                        });
        List<Bookmark> items = fetched.size() > size ? fetched.subList(0, size) : fetched;
        String nextCursor = null;
        if (fetched.size() > size && !items.isEmpty()) {
            Bookmark last = items.get(items.size() - 1);
            nextCursor = new Cursor(last.createdAt(), last.id()).encode();
        }
        return Response.ok(
                        new CursorPage<>(
                                items.stream().map(ServiceSupport::bookmarkResponse).toList(),
                                nextCursor))
                .build();
    }

    public Response createBookmark(BookmarkInput input) {
        Caller caller = requireCaller();
        requireBookmarkVisibility(input.visibility());
        Bookmark created =
                withConnection(
                        connection -> {
                            UUID id = UUID.randomUUID();
                            Instant now = now();
                            return queryOne(
                                            connection,
                                            "insert into bookmarks (id, owner, url, title, notes, tags, visibility, status, created_at, updated_at)"
                                                    + " values (?, ?, ?, ?, ?, ?::text[], ?, 'active', ?, ?) returning *",
                                            params(
                                                    id,
                                                    caller.username(),
                                                    input.url(),
                                                    input.title(),
                                                    input.notes(),
                                                    input.tags(),
                                                    input.visibility(),
                                                    now,
                                                    now),
                                            ServiceSupport::bookmark)
                                    .orElseThrow();
                        });
        return Response.created(URI.create("/api/v1/bookmarks/" + created.id()))
                .entity(bookmarkResponse(created))
                .build();
    }

    public Response getBookmark(String rawId) {
        UUID id = parseUuid(rawId);
        Caller caller = currentCaller();
        Bookmark bookmark =
                withConnection(connection -> findBookmark(connection, id))
                        .orElseThrow(StackverseProblem::notFound);
        if (!bookmark.visibleTo(caller == null ? null : caller.username())) {
            throw StackverseProblem.notFound();
        }
        return Response.ok(bookmarkResponse(bookmark)).build();
    }

    public Response updateBookmark(String rawId, BookmarkInput input) {
        Caller caller = requireCaller();
        UUID id = parseUuid(rawId);
        requireBookmarkVisibility(input.visibility());
        Bookmark updated =
                inTransaction(
                        connection -> {
                            Bookmark bookmark =
                                    queryOne(
                                                    connection,
                                                    "select * from bookmarks where id = ? for update",
                                                    List.of(id),
                                                    ServiceSupport::bookmark)
                                            .orElseThrow(StackverseProblem::notFound);
                            if (!bookmark.owner().equals(caller.username())) {
                                throw StackverseProblem.notFound();
                            }
                            if ("hidden".equals(bookmark.status())
                                    && "public".equals(input.visibility())) {
                                throw StackverseProblem.conflictKey(
                                        "error.bookmark.hidden-publish");
                            }
                            return queryOne(
                                            connection,
                                            "update bookmarks set url = ?, title = ?, notes = ?, tags = ?::text[], visibility = ?, updated_at = ?"
                                                    + " where id = ? returning *",
                                            params(
                                                    input.url(),
                                                    input.title(),
                                                    input.notes(),
                                                    input.tags(),
                                                    input.visibility(),
                                                    now(),
                                                    id),
                                            ServiceSupport::bookmark)
                                    .orElseThrow();
                        });
        return Response.ok(bookmarkResponse(updated)).build();
    }

    private static void requireBookmarkVisibility(String visibility) {
        if (!Set.of("private", "public").contains(visibility)) {
            throw StackverseProblem.badRequest("visibility must be one of: private, public");
        }
    }

    public Response deleteBookmark(String rawId) {
        Caller caller = requireCaller();
        UUID id = parseUuid(rawId);
        withConnection(
                connection -> {
                    Bookmark bookmark =
                            findBookmark(connection, id).orElseThrow(StackverseProblem::notFound);
                    if (!bookmark.owner().equals(caller.username())) {
                        throw StackverseProblem.notFound();
                    }
                    execute(connection, "delete from bookmarks where id = ?", List.of(id));
                    return null;
                });
        return Response.noContent().build();
    }

    public Response listTags() {
        Caller caller = requireCaller();
        List<TagCount> tags =
                withConnection(
                        connection ->
                                query(
                                        connection,
                                        "select tag, count(*) as count from bookmarks, unnest(tags) as tag"
                                                + " where owner = ? group by tag order by count desc, tag",
                                        List.of(caller.username()),
                                        rs ->
                                                new TagCount(
                                                        rs.getString("tag"), rs.getLong("count"))));
        return Response.ok(new TagsResponse(tags)).build();
    }
}

File: backends/scala-http4s/src/main/scala/dev/stackverse/http4s/BookmarkService.scala
package dev.stackverse.http4s

import cats.effect.IO
import io.circe.{Json, JsonObject}
import org.http4s.*

import java.sql.{Connection, SQLException}
import java.time.{Instant, LocalDate, ZoneOffset}
import java.util.UUID
import scala.collection.mutable.ArrayBuffer
import scala.util.Try

trait BookmarkOperations {
  def listBookmarksV1(req: Request[IO]): IO[Response[IO]]
  def listBookmarksV2(req: Request[IO]): IO[Response[IO]]
  def createBookmark(req: Request[IO]): IO[Response[IO]]
  def getBookmark(req: Request[IO], id: String): IO[Response[IO]]
  def updateBookmark(req: Request[IO], id: String): IO[Response[IO]]
  def deleteBookmark(req: Request[IO], id: String): IO[Response[IO]]
  def listTags(req: Request[IO]): IO[Response[IO]]
}

final class BookmarkService(db: Db, auth: AuthService, repository: SharedRepository) extends BookmarkOperations {
  import InputValidation.*
  import RouteSupport.*
  import Wire.*
  import repository.*

  private val TagPattern = "^[a-z0-9-]{1,30}$".r

  override def listBookmarksV1(req: Request[IO]): IO[Response[IO]] = IO.blocking {
    val caller = auth.optional(req)
    val (page, size) = paging(query(req))
    val filters = parseListFilters(query(req))
    val (where, params) = listingWhere(caller, filters)
    val payload = db.withConnection { conn =>
      val rows = db.query(
        conn,
        s"""select * from bookmarks where $where
           |order by created_at desc, id desc
           |limit ? offset ?""".stripMargin,
        params ++ Seq(size, page * size)
      )(Rows.bookmark)
      val total = count(conn, s"select count(*) as count from bookmarks where $where", params)
      pagePayload(rows.map(Responses.bookmark), page, size, total)
    }
    jsonResponse(
      Status.Ok,
      payload,
      "Deprecation" -> "@1782864000",
      "Sunset" -> "Thu, 01 Jul 2027 00:00:00 GMT",
      "Link" -> "</api/v2/bookmarks>; rel=\"successor-version\""
    )
  }

  override def listBookmarksV2(req: Request[IO]): IO[Response[IO]] = IO.blocking {
    val caller = auth.optional(req)
    val (_, size) = paging(query(req))
    val filters = parseListFilters(query(req))
    val cursor = single(query(req), "cursor").map(CursorCodec.decode)
    val (baseWhere, baseParams) = listingWhere(caller, filters)
    val (where, params) = cursor match {
      case Some(value) =>
        (
          s"$baseWhere and (created_at < ? or (created_at = ? and id < ?))",
          baseParams ++ Seq(value.createdAt, value.createdAt, value.id)
        )
      case None => (baseWhere, baseParams)
    }
    val payload = db.withConnection { conn =>
      val fetched = db.query(
        conn,
        s"""select * from bookmarks where $where
           |order by created_at desc, id desc
           |limit ?""".stripMargin,
        params :+ (size + 1)
      )(Rows.bookmark)
      val items = fetched.take(size)
      val nextCursor = if (fetched.size > size && items.nonEmpty) {
        val last = items.last
        Some(CursorCodec.encode(BookmarkCursor(last.createdAt, last.id)))
      } else None
      obj(
        "items" -> Some(Json.arr(items.map(Responses.bookmark)*)),
        "nextCursor" -> nextCursor.map(Json.fromString)
      )
    }
    jsonResponse(Status.Ok, payload)
  }

  override def createBookmark(req: Request[IO]): IO[Response[IO]] =
    for {
      body <- jsonBody(req)
      response <- IO.blocking {
        val caller = auth.requireCaller(req)
        val input = validateBookmarkInput(body)
        val now = Instant.now()
        val id = UUID.randomUUID()
        val row = db.withConnection { conn =>
          db.returning(
            conn,
            """insert into bookmarks (id, owner, url, title, notes, tags, visibility, status, created_at, updated_at)
              |values (?, ?, ?, ?, ?, ?::text[], ?, 'active', ?, ?) returning *""".stripMargin,
            Seq(id, caller.username, input.url, input.title, input.notes, input.tags, input.visibility, now, now)
          )(Rows.bookmark)
        }
        jsonResponse(Status.Created, Responses.bookmark(row), "Location" -> s"/api/v1/bookmarks/$id")
      }
    } yield response

  override def getBookmark(req: Request[IO], id: String): IO[Response[IO]] = IO.blocking {
    val caller = auth.optional(req)
    val uuid = parseUuid(id)
    val row = db.withConnection(conn => findBookmark(conn, uuid))
    val visible = row.exists(bookmark =>
      bookmark.owner == caller
        .map(_.username)
        .orNull || (bookmark.visibility == "public" && bookmark.status == "active")
    )
    if (!visible) throw NotFoundProblem()
    jsonResponse(Status.Ok, Responses.bookmark(row.get))
  }

  override def updateBookmark(req: Request[IO], id: String): IO[Response[IO]] =
    for {
      body <- jsonBody(req)
      response <- IO.blocking {
        val caller = auth.requireCaller(req)
        val uuid = parseUuid(id)
        val input = validateBookmarkInput(body)
        val row = db.transaction { conn =>
          val existing = db.one(conn, "select * from bookmarks where id = ? for update", Seq(uuid))(Rows.bookmark)
          val bookmark = existing.filter(_.owner == caller.username).getOrElse(throw NotFoundProblem())
          if (bookmark.status == "hidden" && input.visibility == "public") {
            throw ConflictProblem(
              "This bookmark was hidden by moderation and cannot be made public.",
              Some("error.bookmark.hidden-publish")
            )
          }
          db.returning(
            conn,
            """update bookmarks set url = ?, title = ?, notes = ?, tags = ?::text[], visibility = ?, updated_at = ?
              |where id = ? returning *""".stripMargin,
            Seq(input.url, input.title, input.notes, input.tags, input.visibility, Instant.now(), uuid)
          )(Rows.bookmark)
        }
        jsonResponse(Status.Ok, Responses.bookmark(row))
      }
    } yield response

  override def deleteBookmark(req: Request[IO], id: String): IO[Response[IO]] = IO.blocking {
    val caller = auth.requireCaller(req)
    val uuid = parseUuid(id)
    db.withConnection { conn =>
      val bookmark = findBookmark(conn, uuid).filter(_.owner == caller.username).getOrElse(throw NotFoundProblem())
      db.execute(conn, "delete from bookmarks where id = ?", Seq(bookmark.id))
    }
    Response[IO](Status.NoContent)
  }

  override def listTags(req: Request[IO]): IO[Response[IO]] = IO.blocking {
    val caller = auth.requireCaller(req)
    val tags = db.withConnection { conn =>
      db.query(
        conn,
        """select tag, count(*) as count
          |from bookmarks, unnest(tags) as tag
          |where owner = ?
          |group by tag
          |order by count desc, tag asc""".stripMargin,
        Seq(caller.username)
      )(rs => Json.obj("tag" -> Json.fromString(rs.getString("tag")), "count" -> Json.fromLong(rs.getLong("count"))))
    }
    jsonResponse(Status.Ok, Json.obj("tags" -> Json.arr(tags*)))
  }

  private def parseListFilters(queryParams: Map[String, Seq[String]]): ListFilters = {
    val q = single(queryParams, "q")
    maxLength(q, 200, "q")
    val visibility = single(queryParams, "visibility")
    if (visibility.exists(value => value != "private" && value != "public"))
      throw BadRequestProblem(s"unknown visibility: ${visibility.get}")
    ListFilters(validateQueryTags(multi(queryParams, "tag")), q, visibility)
  }

  private def validateQueryTags(values: Seq[String]): Seq[String] = {
    val tags = values.map(_.trim.toLowerCase)
    val validator = Validator()
    validator.check(tags.forall(tag => TagPattern.matches(tag)), "tag", "validation.tag.invalid")
    validator.throwIfInvalid()
    tags
  }

  private def listingWhere(caller: Option[Caller], filters: ListFilters): (String, Seq[Any]) = {
    val clauses = ArrayBuffer.empty[String]
    val params = ArrayBuffer.empty[Any]
    if (filters.visibility.contains("public")) {
      clauses += "visibility = 'public' and status = 'active'"
    } else {
      val username = caller.map(_.username).getOrElse(throw UnauthorizedProblem())
      clauses += "owner = ?"
      params += username
      filters.visibility.foreach { value =>
        clauses += "visibility = ?"
        params += value
      }
    }
    if (filters.tags.nonEmpty) {
      clauses += "tags @> ?::text[]"
      params += filters.tags
    }
    filters.q.filter(_.trim.nonEmpty).foreach { value =>
      val pattern = s"%${escapeLike(value)}%"
      clauses += "(title ilike ? escape '\\' or notes ilike ? escape '\\')"
      params += pattern
      params += pattern
    }
    (clauses.mkString(" and "), params.toSeq)
  }

}
